tokenizers/src/models/unigram/trainer.rs

```diff
@@ -293,16 +293,49 @@ impl UnigramTrainer {
         let mut vsum = 0.0;
         let mut freq: Vec<f64> = vec![0.0; pieces.len()];
         let mut inverted: Vec<Vec<usize>> = vec![Vec::new(); pieces.len()];
-        // TODO reparallelize this
-        for (i, (sentence, count)) in sentences.iter().enumerate() {
-            let mut lattice = Lattice::from(sentence, bos_id, eos_id);
-            model.populate_nodes(&mut lattice);
-            vsum += *count as f64;
-            for node_ref in lattice.viterbi() {
-                let id = node_ref.borrow().id;
-                freq[id] += *count as f64;
-                inverted[id].push(i);
-            }
+
+        use rayon::current_num_threads;
+        use rayon::iter::IndexedParallelIterator;
+        use rayon::iter::IntoParallelRefIterator;
+        use rayon::iter::ParallelIterator;
+
+        let chunk_size = std::cmp::max(sentences.len() / current_num_threads(), 1);
+        let collected: Vec<(f64, Vec<f64>, Vec<Vec<usize>>)> = sentences
+            .par_iter()
+            .enumerate()
+            .chunks(chunk_size)
```

Comment:
Hm, Just read that.

I think you really want `par_chunks` and not `par_iter() + chunks()` since I don't expect it to do the same thing.
Doesn't this reallocate a lot ?

You can recover the ids by using this kind of logic:
https://github.com/google/sentencepiece/blob/master/src/unigram_model_trainer.cc#L337

Not sure the cleanest way to handle this in `rayon` ecosystem but `sentences.iter().enumerate().par_chunks` seems natural and like it should work.

Reply:
Tried 28f98c1f6c0a732ae2a867a09b5b2f300e8f3d00 let me know what you think?